controllers/helmrelease_controller.go

```diff
@@ -252,189 +82,135 @@ func (r *HelmReleaseReconciler) SetupWithManager(mgr ctrl.Manager, opts HelmRele
-		if err == nil && hr.Spec.GetTest().Enable {
-			_, testErr := run.Test(hr)
-			testErr = r.handleHelmActionResult(&hr, revision, testErr, "test", v2.TestSuccessCondition, v2.TestSucceededReason, v2.TestFailedReason)
-
-			// Propagate any test error if not marked ignored.
-			if testErr != nil && !remediation.MustIgnoreTestFailures(hr.Spec.GetTest().IgnoreFailures) {
-				testsPassing := meta.GetCondition(hr.Status.Conditions, v2.TestSuccessCondition)
-				v2.SetHelmReleaseCondition(&hr, v2.ReleasedCondition, corev1.ConditionFalse, testsPassing.Reason, testsPassing.Message)
-				err = testErr
-			}
-		}
+	// Reconcile our dependencies
+	if result, err := r.reconcileDependencies(hr); err != nil {
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityInfo,
+			fmt.Sprintf("%s (retrying in %s)", err.Error(), result.RequeueAfter.String()))
+		// We do not return the error here on purpose, as this would
+		// result in a back-off strategy.
+		return result, nil
 	}
 
-	if err != nil {
-		// Increment failure count for deployment action.
-		remediation.IncrementFailureCount(&hr)
-		// Remediate deployment failure if necessary.
-		if !remediation.RetriesExhausted(hr) || remediation.MustRemediateLastFailure() {
-			if util.ReleaseRevision(rel) <= releaseRevision {
-				log.Info(fmt.Sprintf("skipping remediation, no new release revision created"))
-			} else {
-				var remediationErr error
-				switch remediation.GetStrategy() {
-				case v2.RollbackRemediationStrategy:
-					rollbackErr := run.Rollback(hr)
-					remediationErr = r.handleHelmActionResult(&hr, revision, rollbackErr, "rollback",
-						v2.RemediatedCondition, v2.RollbackSucceededReason, v2.RollbackFailedReason)
-				case v2.UninstallRemediationStrategy:
-					uninstallErr := run.Uninstall(hr)
-					remediationErr = r.handleHelmActionResult(&hr, revision, uninstallErr, "uninstall",
-						v2.RemediatedCondition, v2.UninstallSucceededReason, v2.UninstallFailedReason)
-				}
-				if remediationErr != nil {
-					err = remediationErr
-				}
-			}
+	return r.reconcile(ctx, run, hr)
+}
 
-			// Determine release after remediation.
-			rel, observeLastReleaseErr = run.ObserveLastRelease(hr)
-			if observeLastReleaseErr != nil {
-				err = &ConditionError{
-					Reason: v2.GetLastReleaseFailedReason,
-					Err:    errors.New("failed to get last release revision after remediation"),
-				}
-			}
-		}
+func (r *HelmReleaseReconciler) reconcile(ctx context.Context, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	steps := []func(context.Context, *runner.Runner, *v2.HelmRelease) (ctrl.Result, error){
+		r.reconcileRelease,
+		r.reconcileTest,
+		r.reconcileRemediation,
 	}
-
-	hr.Status.LastReleaseRevision = util.ReleaseRevision(rel)
-
-	if err != nil {
-		reason := meta.ReconciliationFailedReason
-		var cerr *ConditionError
-		if errors.As(err, &cerr) {
-			reason = cerr.Reason
+	result := ctrl.Result{}
+	var errs []error
+	for _, step := range steps {
```

Comment:
I like this approach; is it not the case that you want to halt at the first error, though?

Reply:
No, the idea is that there may be introduced logic later that e.g. settles things if your controller crashed mid-run. If you simply run through all steps while requiring that the methods called self-determine if they actually should do anything, this always gives them a chance every reconciliation run to correct state.

Reply:
So would this be for example retries of helm test/rollback/uninstall actions?

Reply:
Exactly, or running the tests without having made a release, etc.

Reply:
With the approach of having separate steps for `reconcileRelease`(or `reconcileUpgrade`/`reconcileInstall`)/`reconcileTest`, how do we accomplish incrementing the install/upgrade failure counts, which depends on knowing whether there was a *new* failure in either the install/upgrade or a resulting (non-ignored) test? I guess we could increment the failure count in `reconcileRelease`(or `reconcileUpgrade`/`reconcileInstall`) upon failure, and then in `reconcileTest` determine the reason(s) why a helm test needs to be run:

1. successful install (maps to `installFailures`)
2. successful upgrade (maps to `upgradeFailures`)
3. https://github.com/fluxcd/flux2/discussions/103

and then upon test failure, only increment a failure count which is mapped to by the calculated reason(s) for having running the test (1. or 2.)

Reply:
Nevermind, looks like this is already implemented 👍